src/advanced_inv.cpp

```diff
@@ -1892,27 +1984,30 @@ bool advanced_inventory::query_charges( aim_location destarea, const advanced_in
             amount = std::min( weightmax, amount );
         }
     }
+    // handle how many of armour type we can equip (max of 2 per type)
+    if(destarea == AIM_WORN) {
+        const auto &id = sitem.items.front()->typeId();
+        amount = MAX_WORN_PER_TYPE - g->u.amount_worn(id);
+    }
 
     // Now we have the final amount. Query if needed (either requested, or when
     // the destination can not hold all items).
-    if( askamount || amount < input_amount ) {
-        // moving several items (not charges!) from ground it currently not implemented.
-        // TODO: implement this properly, see the code where this is called from.
-        if( !by_charges && sitem.area != AIM_INVENTORY && sitem.area != AIM_WORN ) {
-            amount = 1;
-            return true;
-        }
-        std::string popupmsg = _( "How many do you want to move? (0 to cancel)" );
+    if( action == "MOVE_VARIABLE_ITEM" || amount < input_amount ) {
+        const char *msg = _("How many do you want to move? [Have %d] (0 to cancel)");
+        std::string popupmsg = string_format(msg, sitem.stacks);
         // At this point amount contains the maximal amount that the destination can hold.
         if( amount < input_amount ) {
-            popupmsg = string_format( _( "Destination can only hold %d! Move how many? (0 to cancel) " ),
-                                      amount );
+            msg = _("Destination can only hold %d! Move how many? [Have %d] (0 to cancel)");
+            popupmsg = string_format(msg, amount, sitem.stacks);
```

Comment:
The message "How many do you want to move? [Have %d] (0 to cancel)" is not really correct for items counted by charges. It uses `sitem.stacks`, which is always 1 for items counted by charge. Trying to move a stack of "vitamin (20)" allows to enter 10 (half of the stack), but the message says "[Have 1]", which is a bit confusing. It indicates one can only move 0 or 1 item.